src/main/daml/Daml/Finance/Claims/Util/Builders.daml

```diff
@@ -167,23 +167,47 @@ createFxAdjustedPrincipalClaim dateToTime ownerReceives fxRateMultiplier notiona
     prepareAndTagClaims dateToTime [fxLegClaim] "Principal payment"
 -- FIXED_RATE_BOND_REDEMPTION_CLAIM_END
 
--- | Create the claim for a cash-settled, automatically exercised option.
-createEuropeanCashClaim : (Date -> Time) -> Bool -> Decimal -> Observable -> Deliverable -> Date ->
-  Bool -> TaggedClaim
-createEuropeanCashClaim dateToTime isLong strike spot cashInstrument expiryDate
-  isCall =
+-- | Create the claim for a long vanilla option (cash-settled, automatically exercised).
+createVanillaOptionClaim : (Date -> Time) -> Decimal -> Observable -> Deliverable -> Date -> Bool
+  -> C
+createVanillaOptionClaim dateToTime strike spot cashInstrument expiryDate isCall =
   let
     intrinsicValue = if isCall then Observe spot - Const strike
-                             else Const strike - Observe spot
+                               else Const strike - Observe spot
     exercised = scale intrinsicValue $ one cashInstrument
     notExercised = zero
     autoExerciseCondition = if isCall then Lte (Const strike, Observe spot)
-                                    else Lte (Observe spot, Const strike)
-    claimAmount = when (TimeGte expiryDate) $
-                  cond autoExerciseCondition exercised notExercised
+                                      else Lte (Observe spot, Const strike)
+  in
+    when (TimeGte expiryDate) $ cond autoExerciseCondition exercised notExercised
+
+-- | Create the claim for a cash-settled, automatically exercised option (long or short).
+createEuropeanCashClaim : (Date -> Time) -> Bool -> Decimal -> Observable -> Deliverable -> Date ->
+  Bool -> TaggedClaim
+createEuropeanCashClaim dateToTime isLong strike spot cashInstrument expiryDate
+  isCall =
+  let
+    claimAmount = createVanillaOptionClaim dateToTime strike spot cashInstrument expiryDate isCall
     claim = if isLong then claimAmount else give claimAmount
   in
-    prepareAndTagClaims dateToTime [claim] "Option payment"
+    prepareAndTagClaims dateToTime [claim] "European option payment"
+
+-- | Create the claim for a barrier option (automatically exercised, cash-settled).
+createBarrierEuropeanCashClaim : (Date -> Time) -> Bool -> Decimal -> Observable -> Deliverable
+  -> Date -> Bool -> Decimal -> Date -> Bool -> Bool -> TaggedClaim
+createBarrierEuropeanCashClaim dateToTime isLong strike spot cashInstrument
+  expiryDate isCall barrier barrierStartDate isDown isOut =
+  let
+    barrierEvent = if isDown then Lte (Observe spot, Const barrier)
+                             else Lte (Const barrier, Observe spot)
+    vanillaOption = createVanillaOptionClaim dateToTime strike spot cashInstrument expiryDate isCall
+    barrierClaimAmountExclStartDate = if isOut
+      then andList [vanillaOption, when barrierEvent $ give vanillaOption]
+      else when barrierEvent vanillaOption
+    claimAmount = when (at barrierStartDate) barrierClaimAmountExclStartDate
```

Comment:
I'm afraid the expiry of the barrier part doesn't really work at the moment (what happens if the barrier is hit after expiry?)

Reply:
You are right, I added tests for barrier hits after expiry and noticed a case where it did not work as expected. I have fixed the logic now by adding `when (TimeLte expiryDate)`.